File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/image/ImageModule.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.image;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.widget.Toast;
import androidx.annotation.NonNull;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.build.ReactBuildConfig;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.views.imagehelper.ImageSource;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import org.pytorch.rn.core.canvas.ImageData;
import org.pytorch.rn.core.javascript.JSContext;
import org.pytorch.rn.core.utils.FileUtils;

@ReactModule(name = "PyTorchCoreImageModule")
public class ImageModule extends ReactContextBaseJavaModule {

  public static final String TAG = "PTLImageModule";

  public static final String NAME = "PyTorchCoreImageModule";

  private ReactApplicationContext mReactContext;

  public ImageModule(ReactApplicationContext reactContext) {
    super(reactContext);
    mReactContext = reactContext;
  }

  @NonNull
  @Override
  public String getName() {
    return NAME;
  }

  @ReactMethod
  public void release(ReadableMap imageRef, Promise promise) throws Exception {
    JSContext.release(imageRef);
    promise.resolve(null);
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public int getWidth(ReadableMap imageRef) {
    IImage image = JSContext.unwrapObject(imageRef);
    return image.getWidth();
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public int getHeight(ReadableMap imageRef) {
    IImage image = JSContext.unwrapObject(imageRef);
    return image.getHeight();
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public float getNaturalWidth(ReadableMap imageRef) {
    IImage image = JSContext.unwrapObject(imageRef);
    return image.getNaturalWidth();
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public float getNaturalHeight(ReadableMap imageRef) {
    IImage image = JSContext.unwrapObject(imageRef);
    return image.getNaturalHeight();
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public float getPixelDensity(ReadableMap imageRef) {
    IImage image = JSContext.unwrapObject(imageRef);
    return image.getPixelDensity();
  }

  @ReactMethod
  public void scale(ReadableMap imageRef, double sx, double sy, Promise promise) {
    IImage image = JSContext.unwrapObject(imageRef);
    IImage scaledImage = image.scale((float) sx, (float) sy);
    JSContext.NativeJSRef ref = JSContext.wrapObject(scaledImage);
    promise.resolve(ref.getJSRef());
  }

  @ReactMethod
  public void fromURL(String urlString, Promise promise) {
    try {
      URL url = new URL(urlString);
      Bitmap bitmap = BitmapFactory.decodeStream(url.openConnection().getInputStream());
      IImage image = new Image(bitmap);
      JSContext.NativeJSRef ref = JSContext.wrapObject(image);
      promise.resolve(ref.getJSRef());
    } catch (IOException e) {
      promise.reject(e);
    }
    promise.reject(new Error("Could not load image from " + urlString));
  }

  @ReactMethod
  public void fromFile(String filepath, Promise promise) {
    File file = new File(filepath);
    if (file.exists()) {
      Bitmap bitmap = BitmapFactory.decodeFile(filepath);
      IImage image = new Image(bitmap);
      JSContext.NativeJSRef ref = JSContext.wrapObject(image);
      promise.resolve(ref.getJSRef());
    } else {
      promise.reject(new Error("File does not exist " + filepath));
    }
  }

  @ReactMethod
  public void fromBundle(final ReadableMap source, Promise promise) {
    final String uri = source.getString("uri");
    final ImageSource imageSource = new ImageSource(mReactContext, uri);
    if (Uri.EMPTY.equals(imageSource.getUri())) {
      warnImageSource(uri);
    }

    try {
      InputStream inputStream;
      if (imageSource.isResource()) {
        // A uri with no scheme (i.e., `null`) is likely to be a resource or
        // local file. Release mode builds bundle the model file in the APK as
        // a raw resource.
        final int resourceId =
            mReactContext
                .getResources()
                .getIdentifier(imageSource.getSource(), "drawable", mReactContext.getPackageName());
        if (resourceId != 0) {
          inputStream = mReactContext.getResources().openRawResource(resourceId);
        } else {
          // Fall back to the local file system
          inputStream = new FileInputStream(uri);
        }
      } else {
        // Get file path to cache image resource or load image resource from
        // cache if loading from URI fails
        final File targetFile = new File(mReactContext.getCacheDir(), uri);

        // Always try to load image resource from URI to make sure it's always
        // the latest version. Only if fetching the image resource from the URI
        // fails, it will load the cached version (if exists).
        FileUtils.downloadUriToFile(imageSource.getSource(), targetFile);

        inputStream = new FileInputStream(targetFile);
      }

      Bitmap bitmap = BitmapFactory.decodeStream(inputStream);
      IImage image = new Image(bitmap);
      JSContext.NativeJSRef ref = JSContext.wrapObject(image);
      promise.resolve(ref.getJSRef());
    } catch (IOException e) {
      promise.reject(e);
    }
  }

  @ReactMethod
  public void fromImageData(final ReadableMap imageDataRef, final boolean scaled, Promise promise) {
    ImageData imageData = JSContext.unwrapObject(imageDataRef);
    IImage image;
    if (scaled) {
      Bitmap bitmap = imageData.getScaledBitmap();
      image = new Image(bitmap);
    } else {
      // Create a copy of the bitmap to allow developers to independently
      // release the image data and this new image. Without a copy, the image
      // will have an invalid bitmap when the image data is released and vice
      // versa.
      Bitmap bitmap = imageData.getBitmap();
      Bitmap bitmapCopy = bitmap.copy(bitmap.getConfig(), bitmap.isMutable());
      image = new Image(bitmapCopy);
    }
    JSContext.NativeJSRef ref = JSContext.wrapObject(image);
    promise.resolve(ref.getJSRef());
  }

  @ReactMethod
  public void toFile(final ReadableMap imageRef, Promise promise) {
    try {
      IImage image = JSContext.unwrapObject(imageRef);
      Bitmap bitmap = image.getBitmap();
      File cacheDir = mReactContext.getCacheDir();
      File file = File.createTempFile("image", ".png", cacheDir);
      FileOutputStream outputStream = new FileOutputStream(file);
      bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream);
      promise.resolve(file.getAbsolutePath());
    } catch (IOException e) {
      promise.reject(e);
    }
  }

  private void warnImageSource(String uri) {
    if (ReactBuildConfig.DEBUG) {
      Toast.makeText(
              mReactContext,
              "Warning: Image source \"" + uri + "\" doesn't exist",
              Toast.LENGTH_SHORT)
          .show();
    }
  }
}

File: react-native-pytorch-core/ios/Image/ImageModule.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import UIKit

@objc(ImageModule)
public class ImageModule: NSObject {
    @objc
    public func getName() -> String {
        return "PyTorchCoreImageModule"
    }

    @objc(fromURL:resolver:rejecter:)
    public func fromURL(_ urlString: NSString,
                        resolver resolve: RCTPromiseResolveBlock,
                        rejecter reject: RCTPromiseRejectBlock) {
        if let cfURL = CFURLCreateWithString(nil, urlString, nil),
        let imageSource = CGImageSourceCreateWithURL(cfURL, nil),
        let image = CGImageSourceCreateImageAtIndex(imageSource, 0, nil) {
            let bitmapImage = Image(image: image)
            let ref = JSContext.wrapObject(object: bitmapImage).getJSRef()
            resolve(ref)
        } else {
            reject(RCTErrorUnspecified, "Couldn't create image from URL", nil)
        }
    }

    @objc(fromFile:resolver:rejecter:)
    public func fromFile(_ filepath: NSString,
                         resolver resolve: RCTPromiseResolveBlock,
                         rejecter reject: RCTPromiseRejectBlock) {
        let path = filepath as String
        let url = URL(fileURLWithPath: path)
        do {
            let data = try Data(contentsOf: url)
            guard let uiImage = UIImage(data: data), let cgImage = uiImage.cgImage else {
                reject(RCTErrorUnspecified, "Couldn't load image \(path)", nil)
                return
            }
            let image = Image(image: cgImage)
            let ref = JSContext.wrapObject(object: image).getJSRef()
            resolve(ref)
        } catch {
            reject(RCTErrorUnspecified, "Couldn't load file \(path)", nil)
        }
    }

    @objc(fromBundle:resolver:rejecter:)
    public func fromBundle(_ assetImage: NSDictionary,
                           resolver resolve: RCTPromiseResolveBlock,
                           rejecter reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.sync {
            if let dictionary = assetImage as? [AnyHashable: Any] {
                let uiImage = Macros.toUIImage(dictionary)
                if let cgImage = uiImage.cgImage {
                    let bitmapImage = Image(image: cgImage)
                    let ref = JSContext.wrapObject(object: bitmapImage).getJSRef()
                    resolve(ref)
                }
            } else {
                reject(RCTErrorUnspecified, "Couldn't create image from bundle", nil)
            }
        }
    }

    @objc(fromImageData:scaled:resolver:rejecter:)
    public func fromImageData(_ imageDataRef: NSDictionary,
                              scaled: Bool,
                              resolver resolve: RCTPromiseResolveBlock,
                              rejecter reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.sync {
            do {
                let imageData = try JSContextUtils.unwrapObject(imageDataRef, ImageData.self)

                var image: IImage
                if scaled {
                    let bitmap = try imageData.getScaledBitmap()
                    image = Image(image: bitmap)
                } else {
                    let pixelDensity = UIScreen.main.scale
                    image = Image(imageData: imageData, pixelDensity: pixelDensity)
                }

                let ref = JSContext.wrapObject(object: image).getJSRef()
                resolve(ref)
            } catch {
                reject(RCTErrorUnspecified, "Could't create image from image data: \(error)", error)
            }
        }
    }

    @objc(toFile:resolver:rejecter:)
    public func toFile(_ imageRef: NSDictionary,
                       resolver resolve: RCTPromiseResolveBlock,
                       rejecter reject: RCTPromiseRejectBlock) {
        do {
            let image = try JSContextUtils.unwrapObject(imageRef, IImage.self)
            guard let bitmap = image.getBitmap() else {
                print("Could not get bitmap from image")
                return
            }
            let uiImage = UIImage(cgImage: bitmap)
            if let data = uiImage.pngData() {
                let uuidFilename = NSUUID().uuidString
                let filename = getDocumentsDirectory().appendingPathComponent("\(uuidFilename).png")
                try? data.write(to: filename)
                resolve(filename.path)
            }
        } catch {
            reject(RCTErrorUnspecified, "Invalid image reference \(error)", error)
            return
        }
    }

    func getDocumentsDirectory() -> URL {
        let paths = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)
        return paths[0]
    }

    @objc
    public func getWidth(_ imageRef: NSDictionary) -> Any {
        do {
            let image = try JSContextUtils.unwrapObject(imageRef, IImage.self)
            return NSNumber(value: Float(image.getWidth()))
        } catch {
            print("Invalid image reference in getWidth")
            return -1
        }
    }

    @objc
    public func getHeight(_ imageRef: NSDictionary) -> Any {
        do {
            let image = try JSContextUtils.unwrapObject(imageRef, IImage.self)
            return NSNumber(value: Float(image.getHeight()))
        } catch {
            print("Invalid image reference in getHeight")
            return -1
        }
    }

    @objc(scale:sx:sy:resolver:rejecter:)
    public func scale(_ imageRef: NSDictionary,
                      sx: NSNumber,
                      sy: NSNumber,
                      resolver resolve: RCTPromiseResolveBlock,
                      rejecter reject: RCTPromiseRejectBlock) {
        do {
            let image = try JSContextUtils.unwrapObject(imageRef, IImage.self)
            let scaledImage = try image.scale(sx: CGFloat(truncating: sx), sy: CGFloat(truncating: sy))
            let ref = JSContext.wrapObject(object: scaledImage).getJSRef()
            resolve(ref)
        } catch {
            reject(RCTErrorUnspecified, "Invalid image reference in scale: \(error)", error)
        }
    }

    @objc(release:resolver:rejecter:)
    public func release(_ imageRef: NSDictionary,
                        resolver resolve: RCTPromiseResolveBlock,
                        rejecter reject: RCTPromiseRejectBlock) {
        do {
            if let imageRef = imageRef as? [ String: String] {
                try JSContext.release(jsRef: imageRef)
            }
            resolve(nil)
        } catch {
            reject(RCTErrorUnspecified, "Invalid image reference in release: \(error)", error)
        }
    }

    @objc(wrapImage:)
    public static func wrapImage(_ image: UIImage) -> NSString? {
      if let cgImage = image.cgImage {
        let bitmapImage = Image(image: cgImage)
        let ref = JSContext.wrapObject(object: bitmapImage).getJSRef()
        return ref["ID"]! as NSString
      } else {
        return nil
      }
    }
}
